[Federation] Segregate DNS related code to separate controller

shashidharatd opened the pull request:
**What this PR does / why we need it**:
This is the continuation of service controller re-factor work as outlined in #41253
This PR segregates DNS related code from service controller to another controller `service-dns controller` which manages the DNS records on the configured DNS provider.
`service-dns controller` monitors the federated services for the ingress annotations and create/update/delete DNS records accordingly.
`service-dns controller` can be optionally disabled and DNS record management could be done by third party components by monitoring the ingress annotations on federated services. (This would enable something like federation middleware for CoreDNS where federation api server could be used as a backend to CoreDNS eliminating the need for etcd storage.)

**Special notes for your reviewer**:

**Release note**:
```
Federation: A new controller for managing DNS records is introduced which can be optionally disabled to enable third party components to manage DNS records for federated services.
```
cc @kubernetes/sig-federation-pr-reviews

Commit:
Segregate DNS related code to separate controller

Commit:
Auto generated bazel build file

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/45034)
<!-- Reviewable:end -->

shashidharatd commented:
@k8s-bot bazel test this

shashidharatd commented:
@k8s-bot unit test this

quinton-hoole commented on federation/pkg/federation-controller/service/servicecontroller.go:
```diff
@@ -790,17 +699,6 @@ func (s *ServiceController) updateFederatedService(fedService *v1.Service, newLB
 	}
 
-	// Ensure DNS records based on Annotations in federated service for all federated clusters
-	if needUpdate && wantsDNSRecords(fedService) {
```
It's not clear to me where this block of code (and the two removed blocks above) have been moved to?

shashidharatd replied:
@quinton-hoole, We may not need this anymore. We look for service status and update the federated ingress annotation. In case of non-loadbalancer type of service, service status will have empty LoadBalancer field and so we don't update the ingress annotation. is this ok?

shashidharatd commented:
PR rebased. @quinton-hoole, @marun request to PTAL

shashidharatd commented:
/assign @marun

marun commented:
@shashidharatd Have you considered using a shared informer to serve up services to both the service and dns controllers?  My naive assumption is the data requirements are going to be the same between the 2 controllers.

shashidharatd commented:
@marun, its a good suggestion. i will check it out and update. Thanks

shashidharatd commented:
@k8s-bot gce etcd3 e2e test this

shashidharatd commented:
@marun, i gave a shot at using shared informer. Currently shared informer cannot be used by federation controllers since it expects a [clientset.Interface](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/clientset_generated/clientset/clientset.go#L44). here is the [reference](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/informers/informers_generated/externalversions/factory.go#L54). Federation controllers use [federation_clientset.Interface](https://github.com/kubernetes/kubernetes/blob/master/federation/client/clientset_generated/federation_clientset/clientset.go#L31). There is incompatibility as of now.

So my suggestion is to go forward with using shared informer for this PR and optimize it in the future. Do you agree?

This is also related to this issue #41302 which we need to address in future.

marun commented:
@shashidharatd Ok, that's fine.  I didn't realize that it wouldn't be compatible.

marun commented:
Modulo some inline concerns, this looks reasonable.  I also think this extraction of the dns code is a good opportunity for further cleanup that will enable better unit test coverage.  Would you be willing to commit to doing that work in a follow-on PR?

---

Reviewed 6 of 6 files at r1, 1 of 4 files at r2.
Review status: 4 of 7 files reviewed at latest revision, 13 unresolved discussions.

---

*[federation/cmd/federation-controller-manager/app/controllermanager.go, line 146 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjx8FI04tvroAbmjebO:-Kjx8FI04tvroAbmjebP:b-7mzf2d) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/cmd/federation-controller-manager/app/controllermanager.go#L146)):*
> ```Go
> 	}
> 
> 	if controllerEnabled(s.Controllers, serverResources, servicecontroller.DNSControllerName, servicecontroller.RequiredResources, true) {
> ```

The previous behavior initialized the dns provider before the services controller.  Is there a reason this PR doens't maintain that behavior (i.e. failing on dns initialization failure before starting services controller)?

(No action required) Not sure why controllerEnabled accepts the defaultValue parameter if it never varies.

---

*[federation/pkg/federation-controller/service/dns.go, line 1 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxD50bAH0zFR9-GKSF:-KjxD50bAH0zFR9-GKSG:ba2j3s8) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L1)):*
> ```Go
> /*
> ```

Is there a reason to keep this controller in the service package if the controller manager instantiates it separately?  That also prompts the question of whether the dns controller should be initialized if the services controller is disabled (as is proposed in the this revision).  When would that configuration make sense?

---

*[federation/pkg/federation-controller/service/dns.go, line 48 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjx9G609GQZwiX5sZ6b:-Kjx9G609GQZwiX5sZ6c:b-g9jphi) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L48)):*
> ```Go
> 	DNSUserAgentName = "federation-service-dns-controller"
> 
> 	// minDNSTtl is the minimum safe DNS TTL value to use (in seconds).  We use this as the TTL for all DNS records.
> ```

Why is DNS in all caps but TTL is not?

---

*[federation/pkg/federation-controller/service/dns.go, line 61 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjx9S5yhTjgk2trEP8f:-Kjx9S5yhTjgk2trEP8g:b-7mtpza) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L61)):*
> ```Go
> 	zoneName string
> 	zoneID   string
> 	// each federation should be configured with a single zone (e.g. "mycompany.com")
> ```

Should this comment apply to ``dnsZone``?  I'm assuming ``dnsZones`` is caching zones from the dnsaas provider.

---

*[federation/pkg/federation-controller/service/dns.go, line 68 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjx9kTPo_7pTViry_yr:-Kjx9kTPo_7pTViry_ys:bgwussf) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L68)):*
> ```Go
> 	// Informer controller for federated services
> 	serviceController cache.Controller
> 	// Client to federated api server
> ```

I'm assuming this comment should be applied to ``federationClient``? 

---

*[federation/pkg/federation-controller/service/dns.go, line 77 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjx9rT4QJbf86sAqn0Y:-Kjx9rT4QJbf86sAqn0Z:b29omgj) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L77)):*
> ```Go
> 	dns, err := dnsprovider.InitDnsProvider(dnsProvider, dnsProviderConfig)
> 	if err != nil {
> 		glog.Fatalf("DNS provider could not be initialized: %v", err)
> ```

I think this log line should be removed or at the least downgraded from fatal, since the controller manager will fail if it receives an error from this method.

---

*[federation/pkg/federation-controller/service/dns.go, line 93 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxAG_Bi-6HxCHe72DV:-KjxAG_Bi-6HxCHe72DW:b-7ropln) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L93)):*
> ```Go
> 	// Check if configuration provided is valid
> 	if err := d.init(); err != nil {
> 		glog.Fatalf("DNS provider failed to initialize: %v", err)
> ```

ditto

---

*[federation/pkg/federation-controller/service/dns.go, line 133 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxAmKPGV0f_0gAyqv_:-KjxAmKQ9rjz8PwuxOD1:b-ssa3d2) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L133)):*
> ```Go
> 		service := item.(*v1.Service)
> 
> 		ingress, err := ParseFederatedServiceIngress(service)
> ```

(No action required) Ugh, ingress is such an overloaded term.

---

*[federation/pkg/federation-controller/service/dns.go, line 146 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxBCysM_iS0Jzf0Zhl:-KjxB2kXw0TxSqqULxyi:b-1m1ajz) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L146)):*
> ```Go
> }
> 
> func (s *ServiceDNSController) init() error {
> ```

A catch-all function name like ``init`` is often a good indication that a function is doing too much, and that appears to be the case here.  Consider breaking this function down into config validation and zone retrieval to simplify maintenance and testing.  Ideally validation of class invariants would be tested separately from zone handling.

---

*[federation/pkg/federation-controller/service/dns.go, line 344 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxCG38JDEQ4h1hvYUc:-KjxCG38JDEQ4h1hvYUd:buy0zit) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L344)):*
> ```Go
> }
> 
> /* ensureDNSRrsets ensures (idempotently, and with minimum mutations) that all of the DNS resource record sets for dnsName are consistent with endpoints.
> ```

Similarly, why is DNS capitalized but ResourceRecordSets abreviated to Rrsets?

---

*[federation/pkg/federation-controller/service/dns.go, line 362 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxCSW-NxFX6PJvq9nO:-KjxCSW-NxFX6PJvq9nP:beckgjn) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L362)):*
> ```Go
> 			if uplevelCname != "" {
> 				glog.V(4).Infof("Creating CNAME to %q for %q", uplevelCname, dnsName)
> 				newRrset := rrsets.New(dnsName, []string{uplevelCname}, minDNSTtl, rrstype.CNAME)
> ```

Ttl -> TTL (same comment where this appears elsewhere in the file)

---

*[federation/pkg/federation-controller/service/dns.go, line 491 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxCmY4etzFkxftVExH:-KjxCmY4etzFkxftVExI:b6wdvpj) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L491)):*
> ```Go
> 	// dnsNames is the path up the DNS search tree, starting at the leaf
> 	dnsNames := []string{
> 		commonPrefix + "." + zoneNames[0] + "." + regionName + "." + serviceDNSSuffix, // zone level - TODO might need other zone names for multi-zone clusters
> ```

Consider using [strings.Join()](https://golang.org/pkg/strings/#Join) to concatenate with a ``.`` separator.

---

*[federation/pkg/federation-controller/service/servicecontroller.go, line 387 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxDvhxxzrLxDzrm2TQ:-KjxDvhxxzrLxDzrm2TR:b-2ftklw) ([raw file](https://github.com/kubernetes/kubernetes/blob/a481a5bca9d23ac1964c95d81978267b8674bcf5/federation/pkg/federation-controller/service/servicecontroller.go#L387)):*
> ```Go
> }
> 
> func wantsDNSRecords(service *v1.Service) bool {
> ```

How does the new controller avoid processing services that aren't of type load balancer?  I can't seem to find a check for the service type.  Should there be one in the ``worker()`` method?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-:-KjxEKnsrN351xfUj2aI:b-tfxyg3)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
Thanks for the review @marun, have addressed the comments. PTAL. I am for bettering unit test coverage. shall pick it up in follow-up PR.

---

Review status: 4 of 7 files reviewed at latest revision, 13 unresolved discussions.

---

*[federation/cmd/federation-controller-manager/app/controllermanager.go, line 146 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjx8FI04tvroAbmjebO:-KjxdCm9dzhbXaTWBueV:b-6ijjuk) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/cmd/federation-controller-manager/app/controllermanager.go#L146)):*
Actually service controller can function without DNS, so i thought it should be ok to go ahead and initialize Service Controller before validating DNS config parameters. after this PR is merged. DNS controller can optionally be switched off. There is a dependency from DNS controller to service controller, however its not a kind of hard dependency. There will be no issue if DNS controller is enabled and service controller is not enabled. In that case the DNS controller will not receive any service watch events.

In any case i am starting the DNS controller before the service controller now.

> (No action required) Not sure why controllerEnabled accepts the defaultValue parameter if it never varies.

Currently all controllers are by default enabled and i feel thats why we see defaultValue to controllerEnabled func may be not required. But in future may be there will be some controller which by default we want to  disable.

---

*[federation/pkg/federation-controller/service/dns.go, line 1 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxD50bAH0zFR9-GKSF:-KjxgEQvWr2xiSGTLwfV:b63vkn6) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L1)):*
It would be good to separate DNS controller to its own package, but currently i wanted to keep the changes to minimum and make the changes just necessary. We shall segregate the DNS related code to another package subsequently.

> That also prompts the question of whether the dns controller should be initialized if the services controller is disabled (as is proposed in the this revision). When would that configuration make sense?

Yes you are right. DNS controller should not be enabled if service controller is disabled. I have update the code now.

---

*[federation/pkg/federation-controller/service/dns.go, line 48 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjx9G609GQZwiX5sZ6b:-Kjxi0tSboNagnGtd4oI:b-z1mdl6) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L48)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Why is DNS in all caps but TTL is not?
</blockquote></details>

DNS is made all caps, because there were warnings from `go lint` and there were no warnings for Ttl. Anyway let me change Ttl to TTL

---

*[federation/pkg/federation-controller/service/dns.go, line 61 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjx9S5yhTjgk2trEP8f:-Kjxj_S2ApcTSJrkzc4F:b3gwm0b) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L61)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Should this comment apply to ``dnsZone``?  I'm assuming ``dnsZones`` is caching zones from the dnsaas provider.
</blockquote></details>

yes, its mistake carried over from previous code. updated.

---

*[federation/pkg/federation-controller/service/dns.go, line 68 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjx9kTPo_7pTViry_yr:-KjxkD3s9L8PpuiCcXLu:b36aql0) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L68)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

I'm assuming this comment should be applied to ``federationClient``? 
</blockquote></details>

yes, another mistake. updated now

---

*[federation/pkg/federation-controller/service/dns.go, line 77 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjx9rT4QJbf86sAqn0Y:-KjxkdH9qlsm7-Oplr04:b-i7jv70) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L77)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

I think this log line should be removed or at the least downgraded from fatal, since the controller manager will fail if it receives an error from this method.
</blockquote></details>

yes, you are right. changed to error.

---

*[federation/pkg/federation-controller/service/dns.go, line 93 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxAG_Bi-6HxCHe72DV:-KjxklwhQ_Q9OVCFdxJ1:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L93)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Done.

---

*[federation/pkg/federation-controller/service/dns.go, line 133 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxAmKPGV0f_0gAyqv_:-KjxkoZPmRawtNCeUZs9:boypa0g) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L133)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) Ugh, ingress is such an overloaded term.
</blockquote></details>

yes agree. lack of better term. I am very bad at coining terms :). request you to suggest a better one.

---

*[federation/pkg/federation-controller/service/dns.go, line 146 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxBCysM_iS0Jzf0Zhl:-Kjy3ohHP10isxuVuYPI:b-2zt9lm) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L146)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

A catch-all function name like ``init`` is often a good indication that a function is doing too much, and that appears to be the case here.  Consider breaking this function down into config validation and zone retrieval to simplify maintenance and testing.  Ideally validation of class invariants would be tested separately from zone handling.
</blockquote></details>

split the init function according to the suggestion.

---

*[federation/pkg/federation-controller/service/dns.go, line 344 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxCG38JDEQ4h1hvYUc:-KjxlVPwZxjciTeQtE2-:b-73mfxe) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L344)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Similarly, why is DNS capitalized but ResourceRecordSets abreviated to Rrsets?
</blockquote></details>

Again there were no warnings from `golint`. Rrset is used heavily in DNS provider. i think it qualifies to be a word now :). 

---

*[federation/pkg/federation-controller/service/dns.go, line 362 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxCSW-NxFX6PJvq9nO:-KjxmM9jj1Q2Wy1vnNmV:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L362)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Ttl -> TTL (same comment where this appears elsewhere in the file)
</blockquote></details>

Done.

---

*[federation/pkg/federation-controller/service/dns.go, line 491 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxCmY4etzFkxftVExH:-KjxmiloZtEDL9L3RedA:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L491)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Consider using [strings.Join()](https://golang.org/pkg/strings/#Join) to concatenate with a ``.`` separator.
</blockquote></details>

Done.

---

*[federation/pkg/federation-controller/service/servicecontroller.go, line 387 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxDvhxxzrLxDzrm2TQ:-KjxpPT933hz1S3Cu0Cw:b-jbzosi) ([raw file](https://github.com/kubernetes/kubernetes/blob/a481a5bca9d23ac1964c95d81978267b8674bcf5/federation/pkg/federation-controller/service/servicecontroller.go#L387)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

How does the new controller avoid processing services that aren't of type load balancer?  I can't seem to find a check for the service type.  Should there be one in the ``worker()`` method?
</blockquote></details>

Ok. i have added back an old function which used to do this in service controller, but at a different place.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-:-Kjy6xtCfOB7OeywRYr5:bugu2t1)*
<!-- Sent from Reviewable.io -->

marun commented:
Thank you for cleanup.  There appears to be some strangeness introduced in the most recent revision around the ``Run`` method, unless I'm missing something this will break the controller.

---

Reviewed 1 of 4 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[federation/pkg/federation-controller/service/dns.go, line 77 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjx9rT4QJbf86sAqn0Y:-KjyYJFGoj9ItHHS5_Xz:b-jj7xlt) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L77)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

yes, you are right. changed to error.
</blockquote></details>

(No action required) Is it desirable to handle any of the errors in this function vs just returning the ``fmt.Errorf()`` result?

---

*[federation/pkg/federation-controller/service/dns.go, line 133 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxAmKPGV0f_0gAyqv_:-KjyeWsXTx-HsiwBoaG1:bwm55e5) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L133)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

yes agree. lack of better term. I am very bad at coining terms :). request you to suggest a better one.
</blockquote></details>

Can you describe what it is?  I'm not familiar enough with the service controller to suggest a better name. 

---

*[federation/pkg/federation-controller/service/dns.go, line 491 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxCmY4etzFkxftVExH:-KjycSlxd9OMF5d50ILn:b-h5ebin) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L491)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

(No action required) Consider joining the dns names separately from their components rather than inline.

---

*[federation/pkg/federation-controller/service/dns.go, line 75 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjydgMrIZ-TbHj36viR:-KjydgMrIZ-TbHj36viS:b-oeljt) ([raw file](https://github.com/kubernetes/kubernetes/blob/d0144680a368579673c8c5aac5dff496be306843/federation/pkg/federation-controller/service/dns.go#L75)):*
> ```Go
> 	serviceDNSSuffix, zoneName, zoneID string) (*ServiceDNSController, error) {
> 	dns, err := dnsprovider.InitDnsProvider(dnsProvider, dnsProviderConfig)
> 	if err != nil || dns == nil {
> ```

Ah, I see that you've moved the check for dns to here.  I don't think the error handling will be correct, though, since err may be nil and provide no detail to either ``HandleError`` or the caller.

---

*[federation/pkg/federation-controller/service/dns.go, line 88 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjyYjN1yWPirrgdORAZ:-KjyYjN1yWPirrgdORA_:b-urh5jq) ([raw file](https://github.com/kubernetes/kubernetes/blob/d0144680a368579673c8c5aac5dff496be306843/federation/pkg/federation-controller/service/dns.go#L88)):*
> ```Go
> 		workQueue:        workqueue.New(),
> 	}
> 	if err := d.validateConfigParemeters(); err != nil {
> ```

Paremeters -> Parameters (could also just use validateConfig)

---

*[federation/pkg/federation-controller/service/dns.go, line 140 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjyZ7m_j-9da7qW_PYg:-KjyZ7m_j-9da7qW_PYh:b-1yrntn) ([raw file](https://github.com/kubernetes/kubernetes/blob/d0144680a368579673c8c5aac5dff496be306843/federation/pkg/federation-controller/service/dns.go#L140)):*
> ```Go
> func (s *ServiceDNSController) worker() {
> 	for {
> 		func() {
> ```

Why is it desirable to wrap this in a function?  Couldn't a similar result be achieved by using ``continue`` instead of ``return``?  I'm also not sure it makes sense to mix a for loop with ``wait.Until``.  Consider removing the for loop and decreasing the wait interval to achieve a similar result with less overhead.

---

*[federation/pkg/federation-controller/service/dns.go, line 211 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjyc1TAHKNlYVb4iEEv:-Kjyc1TAHKNlYVb4iEEw:beqg5on) ([raw file](https://github.com/kubernetes/kubernetes/blob/d0144680a368579673c8c5aac5dff496be306843/federation/pkg/federation-controller/service/dns.go#L211)):*
> ```Go
> 		s.dnsZone = matchingZones[0]
> 	}
> 	return fmt.Errorf("Multiple matching DNS zones found for %q; please specify zoneID", s.zoneName)
> ```

Am I missing something, or will this statement always be executed (function will always return an error)?

---

*[federation/pkg/federation-controller/service/servicecontroller.go, line 253 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjyd-bs_FJa_BaDXohK:-Kjyd-bs_FJa_BaDXohL:b-wynzoe) ([raw file](https://github.com/kubernetes/kubernetes/blob/d0144680a368579673c8c5aac5dff496be306843/federation/pkg/federation-controller/service/servicecontroller.go#L253)):*
> ```Go
> 
> 	glog.Infof("Starting federation service controller")
> 	defer glog.Infof("Shutting down federation service controller")
> ```

What's to stop this function from starting and shutting down the controller in a single invocation now that the stop channel is not being checked before running the ``defer`` statements?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-:-KjyfT0jZw2ZR0s8GQ-h:bbewyrf)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
I had caused an issue in previous commit which is fixed now and all tests are passing. Also i tried addressing all the comments. PTAL

---

Review status: 5 of 7 files reviewed at latest revision, 6 unresolved discussions.

---

*[federation/pkg/federation-controller/service/dns.go, line 77 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjx9rT4QJbf86sAqn0Y:-KjzrfhJWkN75FZBkkmD:b-8uymun) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L77)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) Is it desirable to handle any of the errors in this function vs just returning the ``fmt.Errorf()`` result?
</blockquote></details>

The error handling is at the source of the error and also caller of this function is not doing `runtime.HandleError`. so i think it should be Ok the way it is.

---

*[federation/pkg/federation-controller/service/dns.go, line 133 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxAmKPGV0f_0gAyqv_:-Kk-Evaj8MnIHroKJFBR:b-34rvup) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L133)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Can you describe what it is?  I'm not familiar enough with the service controller to suggest a better name. 
</blockquote></details>

The federated service holds information about how the service within a cluster can be accessed from outside the cluster in annotation with the type as described [here](https://github.com/kubernetes/kubernetes/blob/master/federation/apis/federation/types.go#L159)

---

*[federation/pkg/federation-controller/service/dns.go, line 491 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxCmY4etzFkxftVExH:-Kk-E-Oe9utCPjy7oSdY:b-rrwakm) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L491)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

(No action required) Consider joining the dns names separately from their components rather than inline.
</blockquote></details>

This is done like this to provide similar readability as earlier code.

---

*[federation/pkg/federation-controller/service/dns.go, line 75 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjydgMrIZ-TbHj36viR:-Kjzqmh6n94sDDVVaOTz:bsfg2kk) ([raw file](https://github.com/kubernetes/kubernetes/blob/d0144680a368579673c8c5aac5dff496be306843/federation/pkg/federation-controller/service/dns.go#L75)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Ah, I see that you've moved the check for dns to here.  I don't think the error handling will be correct, though, since err may be nil and provide no detail to either ``HandleError`` or the caller.
</blockquote></details>

moved it back.

---

*[federation/pkg/federation-controller/service/dns.go, line 88 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjyYjN1yWPirrgdORAZ:-KjzsB9hZwOWhxccFSRg:bo99xc8) ([raw file](https://github.com/kubernetes/kubernetes/blob/d0144680a368579673c8c5aac5dff496be306843/federation/pkg/federation-controller/service/dns.go#L88)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Paremeters -> Parameters (could also just use validateConfig)
</blockquote></details>

mistake, :) fixed now.

---

*[federation/pkg/federation-controller/service/dns.go, line 140 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjyZ7m_j-9da7qW_PYg:-Kk-CVetiWl7searjBRv:b-1dvmxk) ([raw file](https://github.com/kubernetes/kubernetes/blob/d0144680a368579673c8c5aac5dff496be306843/federation/pkg/federation-controller/service/dns.go#L140)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Why is it desirable to wrap this in a function?  Couldn't a similar result be achieved by using ``continue`` instead of ``return``?  I'm also not sure it makes sense to mix a for loop with ``wait.Until``.  Consider removing the for loop and decreasing the wait interval to achieve a similar result with less overhead.
</blockquote></details>

the wrapper function is for defer statement to do workQueue.Done. Also the worker will always be blocked at Get waiting for items. Only when the queue is shutdown the worker will quit the loop, and we had to use wait.Until to restart the worker.

This pattern for worker is quite common and we can find this in few other places within kubernetes code. I have done some update to make this more clear & also updated the serviceController worker to do the same.

---

*[federation/pkg/federation-controller/service/dns.go, line 211 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjyc1TAHKNlYVb4iEEv:-Kk-DeHidqTaysCWa9yA:b-1vzrqp) ([raw file](https://github.com/kubernetes/kubernetes/blob/d0144680a368579673c8c5aac5dff496be306843/federation/pkg/federation-controller/service/dns.go#L211)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Am I missing something, or will this statement always be executed (function will always return an error)?
</blockquote></details>

yeah, i introduced a bug mistakenly. Fixed it now.

---

*[federation/pkg/federation-controller/service/servicecontroller.go, line 253 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-Kjyd-bs_FJa_BaDXohK:-Kk-EJSXJKc9TVKQPLT8:b-4dizn5) ([raw file](https://github.com/kubernetes/kubernetes/blob/d0144680a368579673c8c5aac5dff496be306843/federation/pkg/federation-controller/service/servicecontroller.go#L253)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

What's to stop this function from starting and shutting down the controller in a single invocation now that the stop channel is not being checked before running the ``defer`` statements?
</blockquote></details>

there is <-stopCh at the end of this function.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-:-Kk-FbPJ_DSmmnu_JxA2:brqatmz)*
<!-- Sent from Reviewable.io -->

marun commented:
Thank you for fixing.  Code looks to be in good shape.  My primary concern is avoiding non-trivial nested functions, which are really hard to test effectively.

---

Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[federation/pkg/federation-controller/service/dns.go, line 133 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KjxAmKPGV0f_0gAyqv_:-KkCuTtTh7BinPqky6y9:b-xkk38e) ([raw file](https://github.com/kubernetes/kubernetes/blob/ecee9caa40e5918dc989c5f2dd2002d21ee1f0bd/federation/pkg/federation-controller/service/dns.go#L133)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

The federated service holds information about how the service within a cluster can be accessed from outside the cluster in annotation with the type as described [here](https://github.com/kubernetes/kubernetes/blob/master/federation/apis/federation/types.go#L159)
</blockquote></details>

Ah, so it stems from the LoadBalancerIngress collection.  Not really something to be changed in federation then.

---

*[federation/pkg/federation-controller/service/dns.go, line 139 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KkCvAWVYF0mVq91gzeJ:-KkCvAWVYF0mVq91gzeK:b-42tl1) ([raw file](https://github.com/kubernetes/kubernetes/blob/57662d122bea251c16fe16923ace66e5343813ac/federation/pkg/federation-controller/service/dns.go#L139)):*
> ```Go
> 
> func (s *ServiceDNSController) worker() {
> 	workFunc := func() bool {
> ```

I think this should be a first-class method of ServiceDNSController to ensure that it can be unit tested.

---

*[federation/pkg/federation-controller/service/servicecontroller.go, line 248 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KkCvnLmoJpZq9ICHmQa:-KkCvnLmoJpZq9ICHmQb:b93ewk4) ([raw file](https://github.com/kubernetes/kubernetes/blob/57662d122bea251c16fe16923ace66e5343813ac/federation/pkg/federation-controller/service/servicecontroller.go#L248)):*
> ```Go
> // It's an error to call Run() more than once for a given ServiceController
> // object.
> func (s *ServiceController) Run(workers int, stopCh <-chan struct{}) {
> ```

I think localization of defer statements makes sense, so that addition or removal of a given controller component can be done in one place.  I don't think this change needs to be included in this PR, though.  Consider submitting it separately.

---

*[federation/pkg/federation-controller/service/servicecontroller.go, line 293 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KkCwUyIePjok1osww4y:-KkCwUyIePjok1osww4z:b-cdh2fx) ([raw file](https://github.com/kubernetes/kubernetes/blob/57662d122bea251c16fe16923ace66e5343813ac/federation/pkg/federation-controller/service/servicecontroller.go#L293)):*
> ```Go
> // fedServiceWorker runs a worker thread that just dequeues items, processes them, and marks them done.
> func (s *ServiceController) fedServiceWorker() {
> 	workFunc := func() bool {
> ```

Same comment about making this a first-class function that can be unit tested.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-:-KkCwa8Vj9z3NpdH8S2f:b-kpoa74)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
@k8s-bot kops aws e2e test this

shashidharatd commented:
Review status: 5 of 7 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[federation/pkg/federation-controller/service/dns.go, line 139 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KkCvAWVYF0mVq91gzeJ:-KkE_GPNAWrEuEKj9f1P:b-jsblgw) ([raw file](https://github.com/kubernetes/kubernetes/blob/57662d122bea251c16fe16923ace66e5343813ac/federation/pkg/federation-controller/service/dns.go#L139)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

I think this should be a first-class method of ServiceDNSController to ensure that it can be unit tested.
</blockquote></details>

I have a separate function for workerFunction now which could be tested. is this ok?

---

*[federation/pkg/federation-controller/service/servicecontroller.go, line 248 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KkCvnLmoJpZq9ICHmQa:-KkE_2KTPGkVWhIbxntv:bdsa89t) ([raw file](https://github.com/kubernetes/kubernetes/blob/57662d122bea251c16fe16923ace66e5343813ac/federation/pkg/federation-controller/service/servicecontroller.go#L248)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

I think localization of defer statements makes sense, so that addition or removal of a given controller component can be done in one place.  I don't think this change needs to be included in this PR, though.  Consider submitting it separately.
</blockquote></details>

Okie, will send them in separate PR. preserved old code.

---

*[federation/pkg/federation-controller/service/servicecontroller.go, line 293 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-KkCwUyIePjok1osww4y:-KkE_jUB0B7HrVGVekRL:b-rqxvo7) ([raw file](https://github.com/kubernetes/kubernetes/blob/57662d122bea251c16fe16923ace66e5343813ac/federation/pkg/federation-controller/service/servicecontroller.go#L293)):*
<details><summary><i>Previously, marun (Maru Newby) wrote…</i></summary><blockquote>

Same comment about making this a first-class function that can be unit tested.
</blockquote></details>

ditto.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
@marun, i have addressed your comments. PTAL. Thanks!

marun commented:
/lgtm

I think this is another good step in refactoring the service controller.  I'm hoping future iterations can focus on ensuring good unit test coverage.

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45034#-:-KkHFZyM14Hd7yjqIax_:b9uax9q)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
Thanks @marun for lgtm.
As @quinton-hoole is away for a while, @csbell, can you PTAL and approve this PR. Thanks !

shashidharatd commented:
/assign @csbell

csbell commented:
/lgtm
